File: backend.php
<?php
header('Content-Type: text/html; charset=utf-8');

// Palabras clave asociadas a comandos específicos
$commandKeywords = array(
    "adios" => array("adios", "hasta luego", "nos vemos", "chau"),
    "hola" => array("hola"),
    "ver_alojamientos_populares" => array("populares"),
    "mostrar_alojamientos" => array("todos"),
    "mostrar_venta" => array("venta"),
    "mostrar_renta" => array("renta"),
);

// Obtener el mensaje del usuario del POST
$userInput = $_POST["userInput"];

// Convertir el mensaje a una palabra clave o a una colonia
$keywordOrColony = convert_to_keyword_or_colony($userInput);

// Procesar la consulta del usuario y obtener una respuesta
$prologResult = procesar_consulta($userInput, $keywordOrColony);

// Devolver la respuesta
echo $prologResult;

// Consultas específicas
function procesar_consulta($userInput, $keywordOrColony) {
    if (preg_match('/(características|caracteristicas) (de|tiene) (.+)/i', $userInput, $matches)) {
        return consultar_caracteristicas($matches[3]);
    } elseif (preg_match('/(precio|presupuesto) de (\d+)/i', $userInput, $matches)) {
        return consultar_precio($matches[2]);
    } elseif (preg_match('/(quién|quien) es el propietario de (.+)/i', $userInput, $matches)) {
        return consultar_propietario($matches[2]);
    } elseif (preg_match('/(\d+) cuartos/i', $userInput, $matches)) {
        return consultar_cuartos($matches[1]);
    } elseif (preg_match('/(cómo|como) contacto a (.+)/i', $userInput, $matches)) {
        return contactar_propietario($matches[2]);
    } elseif (preg_match('/(venta|renta) en (.+)/i', $userInput, $matches)) {
        return consultar_por_operacion_y_colonia($matches[1], $matches[2]);
    }
    return prolog_query($keywordOrColony);
}

// Función para manejar consultas sobre operación y colonia
function consultar_por_operacion_y_colonia($operacion, $colonia) {
    $prologResponse = shell_exec('swipl -f backend.pl -g "buscar_alojamientos_por_operacion_y_colonia('.$operacion.', '.$colonia.', Respuesta), write(Respuesta), halt." -t halt');

    if ($prologResponse !== null) {
        $alojamientos = explode(',', $prologResponse);
        $formattedResponse = "Alojamientos en la colonia $colonia:\n";
        foreach ($alojamientos as $alojamiento) {
            $formattedResponse .= "- $alojamiento\n";
        }
        return nl2br($formattedResponse);
    } else {
        return "Parece que no hay alojamientos en $operacion en $colonia. Intenta buscar un alojamiento cerca de tu colonia.";
    }
}

// Función para manejar consultas sobre cómo contactar al propietario de un alojamiento
function contactar_propietario($nombreAlojamiento) {
    $prologResponse = shell_exec('swipl -f backend.pl -g "numero_telefono_propietario(\''.$nombreAlojamiento.'\', Propietario), format(\'Puedes contactar ~w al: ~w\', [\''.$nombreAlojamiento.'\', Propietario]), propietario(Propietario, Telefono), write(Telefono), halt." -t halt');
    if ($prologResponse !== null) {
        return $prologResponse;
    } else {
        return "Parece que no hay algún propietario llamado $nombreAlojamiento. Asegúrate de haber escrito correctamente el nombre del propietario.";
    }
}

function consultar_cuartos($habitaciones) {
    $prologResponse = shell_exec('swipl -f backend.pl -g "format_respuesta_habitaciones('.$habitaciones.', Respuesta), write(Respuesta), halt." -t halt');
    if ($prologResponse !== null) {
        return $prologResponse;
    } else {
        return "Parece que tu mensaje puede haber sido mal escrito o no es del todo claro. ¿Podrías brindar más detalles o aclarar con qué necesitas ayuda?";
    }
}

// Función para manejar consultas sobre características de alojamientos específicos
function consultar_caracteristicas($nombreAlojamiento) {
    $prologResponse = shell_exec('swipl -f backend.pl -g "caracteristicas_alojamiento(\''.$nombreAlojamiento.'\', Caracteristicas), write(Caracteristicas), halt." -t halt');
    if ($prologResponse !== null) {
        $alojamientos = explode(',', $prologResponse);
        $formattedResponse = "Caracteristicas de $nombreAlojamiento:\n";
        foreach ($alojamientos as $alojamiento) {
            $formattedResponse .= "- $alojamiento\n";
        }
        return nl2br($formattedResponse);
    } else {
        return "No reconozco ese alojamiento en mi base de conocimientos puedes verificar si esta correctamente escrito";
    }
}

// Función para manejar consultas sobre precio de alojamientos específicos
function consultar_precio($presupuesto) {
    $prologResponse = shell_exec('swipl -f backend.pl -g "precio_alojamiento_presupuesto('.$presupuesto.', NombreAlojamiento, Precio), format(\'Nombre: ~w, Precio: ~w\n\', [NombreAlojamiento, Precio]), fail." -t halt');
    if ($prologResponse !== null) {
        $alojamientos = explode(',', $prologResponse);
        $formattedResponse = "Alojamientos con precio de $presupuesto:\n";
        foreach ($alojamientos as $alojamiento) {
            $formattedResponse .= " $alojamiento\n";
        }
        return nl2br($formattedResponse);
    } else {
        return "No se ha encontrado ningún departamento disponible con el presupuesto de $presupuesto.";
    }
}

// Función para manejar consultas sobre el propietario de un alojamiento específico
function consultar_propietario($nombreAlojamiento) {
    $prologResponse = shell_exec('swipl -f backend.pl -g "propietario_de_alojamiento(\''.$nombreAlojamiento.'\', Propietario), format(\'El propietario de ~w es ~w.\', [\''.$nombreAlojamiento.'\', Propietario]), halt." -t halt');
    if ($prologResponse !== null) {
        return $prologResponse;
    } else {
        return "Parece que tu mensaje puede haber sido mal escrito o no es del todo claro. ¿Podrías brindar más detalles o aclarar con qué necesitas ayuda?";
    }
}

// Función para procesar la consulta del usuario y obtener una respuesta
function prolog_query($keywordOrColony) {
    if ($keywordOrColony === "hola") {
        return "¡Hola! ¿En qué puedo ayudarte?";
    } elseif ($keywordOrColony === "adios") {
        return "Hasta luego. ¡Que tengas un buen día!";
    }

    if (array_key_exists($keywordOrColony, $GLOBALS['commandKeywords'])) {
        $prologResponse = shell_exec('swipl -f backend.pl -g "bot_response(\''.$keywordOrColony.'\', Response), write(Response), halt." -t halt');
        $prologResponse = trim($prologResponse, '[]');
        if ($prologResponse !== null) {
            $alojamientos = explode(',', $prologResponse);
            $formattedResponse = "Alojamientos encontrados fueron:\n";
            foreach ($alojamientos as $alojamiento) {
                $formattedResponse .= "- $alojamiento\n";
            }
            return nl2br($formattedResponse);
        } else {
            return "Parece que no hay alojamientos disponibles. Intenta buscar un alojamiento cerca de tu colonia.";
        }
    } else {
        $prologResponse = shell_exec('swipl -f backend.pl -g "buscar_alojamientos(\''.$keywordOrColony.'\', Response), write(Response), halt." -t halt');
        $prologResponse = trim($prologResponse, '[]');
        if ($prologResponse !== null) {
            $alojamientos = explode(',', $prologResponse);
            $formattedResponse = "Alojamientos en la colonia $keywordOrColony:\n";
            foreach ($alojamientos as $alojamiento) {
                $formattedResponse .= "- $alojamiento\n";
            }
            return nl2br($formattedResponse);
        } else {
            return "Parece que no hay alojamientos en $keywordOrColony. Intenta buscar un alojamiento cerca de tu colonia.";
        }
    }
}

function convert_to_keyword_or_colony($userInput) {
    $userInput = strtolower($userInput);
    $colonies = array(
        "centro", "bacalar", "colonos", "xlapac", "sisal", "chichimila", "tequila", "santa ana", "san juan",
        "san benito", "santa isabel", "san lucas", "san bernardo", "san francisco", "san jose", "san miguel",
        "santiago", "san pedro", "san juditas", "san antonio", "san gabriel"
    );

    foreach ($colonies as $colony) {
        if (strpos($userInput, $colony) !== false) {
            return $colony;
        }
    }

    foreach ($GLOBALS['commandKeywords'] as $command => $associatedKeywords) {
        foreach ($associatedKeywords as $associatedKeyword) {
            if (strpos($userInput, $associatedKeyword) !== false) {
                return $command;
            }
        }
    }

    return "error";
}
?>

File: backend.pl
:- use_module(library(utf8)).
:- set_prolog_flag(encoding, utf8).
%:- encoding(utf8).
% Definición de alojamiento
operacion(venta).
operacion(renta).

tipodepropiedad(departamento).
tipodepropiedad(casa).

% Definición de colonias
colonia('bacalar').
colonia('colonos').
colonia('xlapac').
colonia('sisal').
colonia('chichimila').
colonia('tequila').
colonia('santa ana').
colonia('san juan').
colonia('san benito').
colonia('santa isabel').
colonia('san lucas').
colonia('san bernardo').
colonia('san francisco').
colonia('san jose').
colonia('san miguel').
colonia('santiago').
colonia('san pedro').
colonia('san juditas').
colonia('san antonio').
colonia('san gabriel').

% Colonias cercanas
colonias_cercanas(colonia('bacalar'), colonia('colonos')).
colonias_cercanas(colonia('colonos'), colonia('bacalar')).
colonias_cercanas(colonia('bacalar'), colonia('xlapac')).
colonias_cercanas(colonia('xlapac'), colonia('bacalar')).
colonias_cercanas(colonia('colonos'), colonia('san francisco')).
colonias_cercanas(colonia('san francisco'), colonia('colonos')).
colonias_cercanas(colonia('xlapac'), colonia('san miguel')).
colonias_cercanas(colonia('san miguel'), colonia('xlapac')).
colonias_cercanas(colonia('sisal'), colonia('santa ana')).
colonias_cercanas(colonia('santa ana'), colonia('sisal')).
colonias_cercanas(colonia('san juan'), colonia('san benito')).
colonias_cercanas(colonia('san benito'), colonia('san juan')).
colonias_cercanas(colonia('san lucas'), colonia('san bernardo')).
colonias_cercanas(colonia('san bernardo'), colonia('san lucas')).
colonias_cercanas(colonia('tequila'), colonia('santa isabel')).
colonias_cercanas(colonia('santa isabel'), colonia('tequila')).
colonias_cercanas(colonia('santiago'), colonia('san pedro')).
colonias_cercanas(colonia('san pedro'), colonia('santiago')).
colonias_cercanas(colonia('san juditas'), colonia('san antonio')).
colonias_cercanas(colonia('san antonio'), colonia('san juditas')).

% Propietarios
propietario('Luisa Pérez', 9988774444).
propietario('Jorge Mendoza', 9990001111).
propietario('Marco Cervera', 9990000001).
propietario('Ana Canul', 9990000002).
propietario('Ana Medina', 9990000003).
propietario('David Torres', 9990000004).
propietario('Yahir Ademar', 9990000005).
propietario('Carlos Poot', 9990000006).
propietario('Amy Poot', 9990000007).
propietario('Jafet Cocom', 9990000008).
propietario('Edgar Yam', 9990000009).
propietario('Manuel Masa', 9990000010).
propietario('Daniel Dzib', 9990000011).
propietario('Sandra Romer', 9990000012).
propietario('Lucia Ruiz', 9990000013).
propietario('Roberto Flores', 9990000014).
propietario('Elena Romero', 9990000015).
propietario('Javier Carona', 9990000016).
propietario('Guadalupe Vargas', 9990000017).

% Relaciones de es_propietario
es_propietario('Marco Cervera', 'Departamentos Cerveras').
es_propietario('Ana Canul', 'Casa Girasol').
es_propietario('Ana Medina', 'Casa Anny').
es_propietario('David Torres', 'Torres del Anochecer').
es_propietario('Yahir Ademar', 'Casa Bonita').
es_propietario('Carlos Poot', 'Casa del Parque').
es_propietario('Amy Poot', 'Departamento Amy').
es_propietario('Jafet Cocom', 'Casa Cocom').
es_propietario('Edgar Yam', 'Chalet del Atardecer').
es_propietario('Manuel Masa', 'Casa Encantada').
es_propietario('Daniel Dzib', 'Quinta Primavera').
es_propietario('Sandra Romer', 'Bandidas').
es_propietario('Lucia Ruiz', 'Casita Acogedora').
es_propietario('Roberto Flores', 'Suspiro').
es_propietario('Elena Romero', 'Hogar Dulce Hogar').
es_propietario('Javier Carona', 'Encantos').
es_propietario('Guadalupe Vargas', 'Casa del Sueño').

% Alojamientos
alojamiento('Departamentos Cerveras', tipodepropiedad(departamento), operacion(venta), colonia('bacalar'), 2, 2, valladolid, 1400, disponible).
alojamiento('Casa Girasol', tipodepropiedad(casa), operacion(renta), colonia('bacalar'), 2, 4, valladolid, 2400, disponible).
alojamiento('Casa Anny', tipodepropiedad(casa), operacion(renta), colonia('bacalar'), 1, 2, valladolid, 2400, disponible).
alojamiento('Torres del Anochecer', tipodepropiedad(departamento), operacion(renta), colonia('xlapac'), 1, 3, valladolid, 1400, disponible).
alojamiento('Casa Bonita', tipodepropiedad(casa), operacion(venta), colonia('bacalar'), 1, 5, valladolid, 1400, disponible).
alojamiento('Casa del Parque', tipodepropiedad(casa), operacion(renta), colonia('sisal'), 1, 4, valladolid, 2400, disponible).
alojamiento('Departamento Amy', tipodepropiedad(departamento), operacion(venta), colonia('santa ana'), 1, 2, valladolid, 2400, disponible).
alojamiento('Casa Cocom', tipodepropiedad(casa), operacion(renta), colonia('san juan'), 2, 5, valladolid, 1400, disponible).
alojamiento('Chalet del Atardecer', tipodepropiedad(casa), operacion(renta), colonia('san benito'), 1, 3, valladolid, 2400, disponible).
alojamiento('Apartementos del Amancer', tipodepropiedad(departamento), operacion(venta), colonia('santa isabel'), 1, 5, valladolid, 1400, disponible).
alojamiento('Casa Encantada', tipodepropiedad(casa), operacion(renta), colonia('san lucas'), 1, 4, valladolid, 2400, disponible).
alojamiento('Quinta Primavera', tipodepropiedad(casa), operacion(renta), colonia('san bernardo'), 1, 3, valladolid, 2400, disponible).
alojamiento('Bandidas', tipodepropiedad(departamento), operacion(venta), colonia('san francisco'), 1, 5, valladolid, 1400, disponible).
alojamiento('Casita Acogedora', tipodepropiedad(casa), operacion(renta), colonia('san jose'), 1, 4, valladolid, 2400, disponible).
alojamiento('Suspiro', tipodepropiedad(departamento), operacion(venta), colonia('san miguel'), 1, 2, valladolid, 2400, disponible).
alojamiento('Hogar Dulce Hogar', tipodepropiedad(casa), operacion(renta), colonia('santiago'), 1, 3, valladolid, 2400, disponible).
alojamiento('Encantos', tipodepropiedad(departamento), operacion(venta), colonia('san pedro'), 1, 5, valladolid, 1400, disponible).
alojamiento('Casa del Sueño', tipodepropiedad(casa), operacion(renta), colonia('san juan'), 1, 4, valladolid, 2400, disponible).


% Regla para encontrar todos los alojamientos
find_all_alojamientos(Alojamientos) :-
    findall(Nombre, alojamiento(Nombre, _, _, _, _, _, _, _, _), Alojamientos).

%%%%%
% Encuentra alojamientos en la colonia especificada
find_alojamientos_por_colonia(Colonia, Alojamientos) :-
    findall(Nombre, alojamiento(Nombre, _, _, colonia(Colonia), _, _, _, _, _), Alojamientos).

% Encuentra alojamientos en colonias cercanas a la colonia especificada
find_alojamientos_por_colonias_cercanas(Colonia, Alojamientos) :-
    findall(Nombre, 
        (colonias_cercanas(colonia(Colonia), colonia(ColoniaCercana)),
        alojamiento(Nombre, _, _, colonia(ColoniaCercana), _, _, _, _, _)), 
        Alojamientos).

% Combina los resultados de ambas búsquedas y elimina duplicados
buscar_alojamientos(Colonia, AlojamientosUnicos) :-
    find_alojamientos_por_colonia(Colonia, AlojamientosColonia),
    find_alojamientos_por_colonias_cercanas(Colonia, AlojamientosCercanas),
    append(AlojamientosColonia, AlojamientosCercanas, AlojamientosCombinados),
    sort(AlojamientosCombinados, AlojamientosUnicos).
%%%%%


% Regla para encontrar alojamientos populares por estrellas (4 estrellas o más)
find_alojamientos_populares(Alojamientos) :-
    findall(Nombre, (alojamiento(Nombre, _, _, _, _, Estrellas, _, _, _), Estrellas >= 4), Alojamientos).

% Regla para formatear la respuesta como lista
format_respuesta(Alojamientos, Alojamientos).

% Regla para responder a mensajes específicos
% Regla para responder a mensajes específicos
bot_response(Mensaje, Respuesta) :-
    (   Mensaje = 'hola' ->
        Respuesta = '¡Hola! ¿En qué puedo ayudarte?'
    ;   Mensaje = 'mostrar_alojamientos' ->
        find_all_alojamientos(Alojamientos),
        format_respuesta(Alojamientos, Respuesta)
    ;   Mensaje = 'ver_alojamientos_populares' ->
        find_alojamientos_populares(Alojamientos),
        format_respuesta(Alojamientos, Respuesta)
    ;   Mensaje = 'mostrar_venta' ->
        find_alojamientos_venta(Alojamientos),
        format_respuesta(Alojamientos, Respuesta)
    ;   Mensaje = 'mostrar_renta' ->
        find_alojamientos_renta(Alojamientos),
        format_respuesta(Alojamientos, Respuesta)
    ;   Mensaje = 'adios' ->
        Respuesta = '¡Adiós! Espero verte pronto.'
    ;   Mensaje = 'error' ->
        Respuesta = 'Lo siento, no entendí tu consulta. Por favor, intenta de nuevo.'
    ).

% Regla para encontrar alojamientos en venta
find_alojamientos_venta(Alojamientos) :-
    findall(Nombre, alojamiento(Nombre, _, operacion(venta), _, _, _, _, _, _), Alojamientos).

% Regla para encontrar alojamientos en renta
find_alojamientos_renta(Alojamientos) :-
    findall(Nombre, alojamiento(Nombre, _, operacion(renta), _, _, _, _, _, _), Alojamientos).


% Regla para encontrar las características de un alojamiento por su nombre
caracteristicas_alojamiento(NombreAlojamiento, Caracteristicas) :-
    alojamiento(NombreAlojamiento, tipodepropiedad(Tipo), operacion(Operacion), colonia(Colonia), Habitaciones, Estrellas, Ciudad, Precio, Disponible),
    format(atom(Caracteristicas), 'Nombre: ~w, Tipo de propiedad: ~w, Operación: ~w, Ciudad: ~w, Número de habitaciones: ~w, Calificación: ~w estrellas, Ciudad: ~w, Precio: ~w, Estado: ~w\n', 
           [NombreAlojamiento, Tipo, Operacion, Colonia, Habitaciones, Estrellas, Ciudad, Precio, Disponible]).

% Regla para encontrar alojamientos cuyo precio sea igual o menor que el presupuesto dado
precio_alojamiento_presupuesto(Presupuesto, NombreAlojamiento, Precio) :-
    alojamiento(NombreAlojamiento, _, _, _, _, _, _, Precio, _),
    Precio =< Presupuesto.

% Regla para encontrar el propietario de un alojamiento dado su nombre
propietario_de_alojamiento(NombreAlojamiento, Propietario) :-
    es_propietario(Propietario, NombreAlojamiento).

% Regla para formatear la respuesta sobre el propietario
format_respuesta_propietario(NombreAlojamiento, Respuesta) :-
    (   propietario_de_alojamiento(NombreAlojamiento, Propietario) ->
        format(atom(Respuesta), 'El propietario de ~w es ~w.', [NombreAlojamiento, Propietario])
    ;   format(atom(Respuesta), 'No se encontró el propietario de ~w.', [NombreAlojamiento])
    ).

% Regla para encontrar alojamientos con un número específico de habitaciones
num_habitaciones(NUM_HAB, NombreAlojamiento) :-
    alojamiento(NombreAlojamiento, _, _, _, NUM_HAB, _, _, _, _).

% Regla para formatear la respuesta sobre el número de habitaciones
format_respuesta_habitaciones(NUM_HAB, Respuesta) :-
    findall(Nombre, num_habitaciones(NUM_HAB, Nombre), Alojamientos),
    atomic_list_concat(Alojamientos, ', ', ListaAlojamientos),
    format(atom(Respuesta), 'Alojamientos con ~w habitaciones: ~w', [NUM_HAB, ListaAlojamientos]).

numero_telefono_propietario(NombrePropietario, NumeroTelefono) :-
    propietario(NombrePropietario, NumeroTelefono).

% Regla para buscar alojamientos por operación y colonia
buscar_alojamientos_por_operacion_y_colonia(Operacion, Colonia, Respuesta) :-
    findall(NombreAlojamiento, 
            (alojamiento(NombreAlojamiento, _, operacion(Operacion), colonia(Colonia), _, _, _, _, _)), 
            Alojamientos),
    atomic_list_concat(Alojamientos, ', ', Respuesta).
